controllers/v1beta1/build_qoshandler.go

```diff
@@ -52,18 +51,42 @@ func (r *LagoonBuildReconciler) whichBuildNext(ctx context.Context, opLog logr.L
 	if err := r.List(ctx, runningBuilds, listOption); err != nil {
 		return fmt.Errorf("Unable to list builds in the cluster, there may be none or something went wrong: %v", err)
 	}
+	buildsToStart := r.BuildQoS.MaxBuilds - len(runningBuilds.Items)
 	if len(runningBuilds.Items) >= r.BuildQoS.MaxBuilds {
 		// if the maximum number of builds is hit, then drop out and try again next time
 		if r.EnableDebug {
 			opLog.Info(fmt.Sprintf("Currently %v running builds, no room for new builds to be started", len(runningBuilds.Items)))
 		}
+		go r.processQueue(ctx, opLog, buildsToStart, true)
 		return nil
 	}
-	buildsToStart := r.BuildQoS.MaxBuilds - len(runningBuilds.Items)
 	if buildsToStart > 0 {
 		opLog.Info(fmt.Sprintf("Currently %v running builds, room for %v builds to be started", len(runningBuilds.Items), buildsToStart))
 		// if there are any free slots to start a build, do that here
-		listOption = (&client.ListOptions{}).ApplyOptions([]client.ListOption{
+		go r.processQueue(ctx, opLog, buildsToStart, false)
+	}
+	return nil
+}
+
+var runningProcessQueue bool
```

Comment:
This seems a prime candidate for replacing with a mutex and a defer to clear.

I wonder, does this ever potentially cause issues as it's currently implemented?
I see that we set this to true at the beginning of the process, but then in some cases the process returns without setting it to false again - how does this recover? (does it matter?)

Reply:
Yeah, might have missed some resets in there. Will check.

I'm happy to look at replacing it with a mutex with defer, but previous attempts lead the process to just be delayed execution, not prevention of execution